polars/src/utils.rs

```diff
@@ -62,6 +62,47 @@ pub fn get_iter_capacity<T, I: Iterator<Item = T>>(iter: &I) -> usize {
     }
 }
 
+/// An iterator that iterates an unknown at compile time number
+/// of iterators simultaneously.
+///
+/// IMPORTANT: It differs from `std::iter::Zip` in the return type
+/// of `next`. It returns a `Vec` instead of a `tuple`, which implies
+/// that the result is non-copiable anymore.
+pub struct DynamicZip<I>
+where
+    I: Iterator,
+{
+    iterators: Vec<I>,
+}
+
+impl<I, T> Iterator for DynamicZip<I>
+where
+    I: Iterator<Item = T>,
+{
+    type Item = Vec<T>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        self.iterators.iter_mut().map(|iter| iter.next()).collect()
```

Comment:
Every iteration is a new heap allocation with this iterator. Seems good for a dynamic number of keys, but let's keep the old iterators for up to 5 keys.

Reply:
Yes I agree with you, it is better to keep the old way to avoid allocation for a small set of keys, and the new way just for dynamic grouping.

I propose to expand from 5 to 12 using the old way, as 12 is the maximum size for a tuple to implement the Copy trait, as explained in the [docs](https://doc.rust-lang.org/std/primitive.tuple.html#trait-implementations-1), and after that point to implement this solution.

Do you agree with this or do you prefer just 5?

Reply:
Yes 12 is good. The 5 was just some arbitrary chosen number.

Reply:
I have the solution almost ready I am preparing the test to catch the corner cases.